service/qos/service_level_controller.cc

```diff
@@ -86,15 +86,30 @@ void service_level_controller::set_distributed_data_accessor(service_level_distr
     }
 }
 
-future<> service_level_controller::stop() {
-    // unregister from the service level distributed data accessor.
-    _sl_data_accessor = nullptr;
-    if (this_shard_id() == global_controller) {
-        // abort the loop of the distributed data checking if it is running
-        _global_controller_db->dist_data_update_aborter.request_abort();
-        _global_controller_db->notifications_serializer.broken();
+future<> service_level_controller::drain() {
+    if (this_shard_id() != global_controller) {
+        return make_ready_future();
+    }
+    if (_drained) {
+        return make_ready_future<>();
     }
-    return std::exchange(_distributed_data_updater, make_ready_future<>());
+    _drained = true;
```

Comment:
Could two drains be called in parallel? In that case the second will return early.

If that's a problem, we can use a shared_future to indicate when draining is complete to all callers.

Reply:
Currently the only user of `service_level_controller::drain` is `service_level_controller::stop`, which is only called once on shutdown, so two drains would not be called in parallel.

Reply:
Ok. So I don't see why you needed to extract drain().

Reply:
It turned out I didn't. It's a byproduct of earlier iterations, but it's good practice anyway (because the update process can have virtual executors and depend on multiple services), so I let it stay. I can also remove it from this series if requested.

Reply:
(please let me know if it *is* requested)

Reply:
No, it's fine as is.

Reply:
I don't think that can happen, yet it looks like if it will `std::exchange(_global_controller_db->distributed_data_update, make_ready_future<>())` it would be idempotent and won't need to add the `_drained` flag at all.

Reply:
As long as `abort_source::request_abort()` and `semaphore::broken()` are also idempotent, but it looks like they are. I'll get rid of the flag.

Reply:
Oh wait - semaphore::broken might be, but request_abort isn't, it would have to be wrapped in a check. Still, it's easy.

Reply:
You're right about `request_abort`, sorry.  And indeed `abort_requested()` can come to the rescue.